src/stats/hmm.rs

```diff
@@ -0,0 +1,690 @@
+            display(
+                "inferred from A: N_0={}, N_1={} (must be equal), from B: N={}, M={}, from \
+                pi: N={}", an0, an1, bn, bm, pin)
+        }
+    }
+}
+
+custom_derive! {
+    /// A newtype for HMM states.
+    #[derive(
+        NewtypeFrom,
+        NewtypeDeref,
+        PartialEq,
+        Copy,
+        Clone,
+        Debug
+    )]
+    // #[derive(Serialize, Deserialize)]
+    pub struct State(pub usize);
+}
+
+/// An iterator iterating states.
+pub struct StateIter {
+    nxt: usize,
+    max: usize,
+}
+
+impl Iterator for StateIter {
+    type Item = State;
+
+    fn next(&mut self) -> Option<State> {
+        if self.nxt < self.max {
+            let cur = self.nxt;
+            self.nxt += 1;
+            Some(State(cur))
+        } else {
+            None
+        }
+    }
+}
+
+/// A transition between two states.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct StateTransition {
+    /// Source of the transition.
+    pub src: State,
+    /// Destination of the transition.
+    pub dst: State,
+}
+
+impl StateTransition {
+    /// Constructor.
+    pub fn new(src: State, dst: State) -> Self {
+        Self { src, dst }
+    }
+}
+
+/// Iterator over all state transitions.
+pub struct StateTransitionIter {
+    nxt_a: usize,
+    nxt_b: usize,
+    max: usize,
+}
+
+impl Iterator for StateTransitionIter {
+    type Item = StateTransition;
+
+    fn next(&mut self) -> Option<StateTransition> {
+        let cur_b = self.nxt_b;
+        let cur_a = self.nxt_a;
+        if self.nxt_b < self.max {
+            self.nxt_b += 1;
+            Some(StateTransition::new(State(cur_a), State(cur_b)))
+        } else if self.nxt_a < self.max {
+            self.nxt_b = 0;
+            self.nxt_a += 1;
+            Some(StateTransition::new(State(cur_a), State(cur_b)))
+        } else {
+            None
+        }
+    }
+}
+
+/// A trait for Hidden Markov Models (HMM) with generic observations.
+///
+/// An HMM is characterized by a state of sets, transition probabilities, and some output.
+/// The current interface allows for both discrete and continuous observations.
+pub trait Model<Observation> {
+    /// The number of states in the model.
+    fn num_states(&self) -> usize;
+
+    /// Return iterator over the states of an HMM.
+    fn states(&self) -> StateIter;
+
+    /// Returns an iterator of all transitions.
+    fn transitions(&self) -> StateTransitionIter;
+
+    /// Transition probability between two states `from` and `to`.
+    fn transition_prob(&self, from: State, to: State) -> LogProb;
+
+    /// Initial probability given the HMM `state`.
+    fn initial_prob(&self, state: State) -> LogProb;
+
+    /// Probability for the given observation in the given state.
+    fn observation_prob(&self, state: State, observation: &Observation) -> LogProb;
+}
+
+/// Compute the probability Viterbi matrix and the pointers to the origin.
+fn viterbi_matrices<Observation>(
+    hmm: &Model<Observation>,
```

Comment:
If you use a trait object here, rust will have to dispatch dynamically whenever hmm is accessed. I think it might be faster to use generics, because then the compiler can do all sorts of optimizations like inlining. Same for other functions.

Reply:
It happens simply when you directly refer to a trait, like with &Model.

Reply:
Instead, write something like ` <M: Model>(model: &M)`